[custom_op] register_autograd supports non-tensor kwargonly-args

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124852
* __->__ #124806
* #124805
* #124637

The user does not need to return gradients for these args.

We also change how setup_context works to adapt to kwargonly-args. If
the user's op has no kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, output)`: we require that the
arguments have the same names.

If the user's op has kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, keyword_only_inputs, output)`.
We require that the arguments have the same names.

Test Plan:
- new tests

zou3519 committed:
[custom_op] register_autograd supports non-tensor kwargonly-args

The user does not need to return gradients for these args.

We also change how setup_context works to adapt to kwargonly-args. If
the user's op has no kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, output)`: we require that the
arguments have the same names.

If the user's op has kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, inputs_keyword_only, output)`.
We require that the arguments have the same names.

Test Plan:
- new tests

[ghstack-poisoned]

albanD commented on torch/_library/autograd.py:
```diff
@@ -60,13 +70,15 @@ def backward(ctx, *grads):
         Generated = supports_tensorlist(Generated)
 
-    def autograd_impl(keyset, *args):
+    def autograd_impl(keyset, *args, **kwargs):
```
Should we rename this to args, kwargs_only with a comment saying that non-kwargs-only arguments are moved to be positional by the dispatcher logic already?

zou3519 replied:
Yup, can do

zou3519 commented on torch/_library/autograd.py:
```diff
@@ -22,16 +22,26 @@ def make_autograd_impl(op: _ops.OpOverload, info: InfoProtocol) -> Callable:
     name: str = f"GeneratedBackwardFor_{op._namespace}_{op._opname}_{op._overloadname}"
 
     saved_keyset = None
+    saved_kwargs = None
```
These will change ctx.needs_input_grad (e.g. ctx.needs_input_grad[-1]), so I didn't pass them into the forward

zou3519 committed:
Update on "[custom_op] register_autograd supports non-tensor kwargonly-args"


The user does not need to return gradients for these args.

We also change how setup_context works to adapt to kwargonly-args. If
the user's op has no kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, output)`: we require that the
arguments have the same names.

If the user's op has kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, keyword_only_inputs, output)`.
We require that the arguments have the same names.

Test Plan:
- new tests

[ghstack-poisoned]

albanD reviewed (commented):
Ok!
Lint needs fixing, SGTM otherwise!

zou3519 committed:
Update on "[custom_op] register_autograd supports non-tensor kwargonly-args"


The user does not need to return gradients for these args.

We also change how setup_context works to adapt to kwargonly-args. If
the user's op has no kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, output)`: we require that the
arguments have the same names.

If the user's op has kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, keyword_only_inputs, output)`.
We require that the arguments have the same names.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 committed:
Update on "[custom_op] register_autograd supports non-tensor kwargonly-args"


The user does not need to return gradients for these args.

We also change how setup_context works to adapt to kwargonly-args. If
the user's op has no kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, output)`: we require that the
arguments have the same names.

If the user's op has kwonly-args, then their setup_context function must
look like `setup_context(ctx, inputs, keyword_only_inputs, output)`.
We require that the arguments have the same names.

Test Plan:
- new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
